File: client/lib/features/events/data/models/event_model.dart
import '../../domain/entities/event.dart';

class EventModel extends Event {
  const EventModel({
    required super.id,
    required super.eventName,
    required super.eventDescription,
    required super.eventLocation,
    required super.lat,
    required super.lng,
    required super.eventStart,
    required super.eventEnd,
    required super.eventTicketPrice,
    required super.eventType,
    required super.eventOrganizer,
    required super.eventCapacity,
    required super.eventTags,
    required super.images,
    super.contactInfo,
    required super.totalAttendees,
    required super.totalLikes,
    required super.isActive,
    required super.isFeatured,
    required super.createdAt,
    required super.updatedAt,
  });

  factory EventModel.fromJson(Map<String, dynamic> json) {
    return EventModel(
      id: json['id'] as String,
      eventName: json['event_name'] as String,
      eventDescription: json['event_description'] as String,
      eventLocation: json['event_location'] as String,
      lat: (json['lat'] as num).toDouble(),
      lng: (json['lng'] as num).toDouble(),
      eventStart: DateTime.parse(json['event_start'] as String),
      eventEnd: DateTime.parse(json['event_end'] as String),
      eventTicketPrice: double.parse(json['event_ticket_price'].toString()),
      eventType: json['event_type'] as String,
      eventOrganizer: json['event_organizer'] as String,
      eventCapacity: json['event_capacity'] as int,
      eventTags: (json['event_tags'] as List<dynamic>)
          .map((e) => e.toString())
          .toList(),
      images: (json['images'] as List<dynamic>)
          .map((e) => e.toString())
          .toList(),
      contactInfo: json['contact_info'] as String?,
      totalAttendees: json['total_attendees'] as int,
      totalLikes: json['total_likes'] as int,
      isActive: json['is_active'] as bool,
      isFeatured: json['is_featured'] as bool,
      createdAt: DateTime.parse(json['createdAt'] as String),
      updatedAt: DateTime.parse(json['updatedAt'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'event_name': eventName,
      'event_description': eventDescription,
      'event_location': eventLocation,
      'lat': lat,
      'lng': lng,
      'event_start': eventStart.toIso8601String(),
      'event_end': eventEnd.toIso8601String(),
      'event_ticket_price': eventTicketPrice,
      'event_type': eventType,
      'event_organizer': eventOrganizer,
      'event_capacity': eventCapacity,
      'event_tags': eventTags,
      'images': images,
      'contact_info': contactInfo,
      'total_attendees': totalAttendees,
      'total_likes': totalLikes,
      'is_active': isActive,
      'is_featured': isFeatured,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }
}

File: server/models/event.model.js
import {DataTypes} from 'sequelize';
import sequelize from '../database/db.js';

const Event = sequelize.define('Event', {
    id: {
        type: DataTypes.UUID,
        defaultValue: DataTypes.UUIDV4,
        primaryKey: true,
    },
    event_name: {
        type: DataTypes.STRING(255),
        allowNull: false,
        index: true
    },
    event_description: {
        type: DataTypes.TEXT,
        allowNull: true,
    },
    event_location: {
        type: DataTypes.STRING(255),
        allowNull: true,
    },
    lat: {
        type: DataTypes.FLOAT,
        allowNull: true,
    },
    lng: {
        type: DataTypes.FLOAT,
        allowNull: true,
    },
    geom: {
        type: DataTypes.GEOMETRY('POINT'),
        allowNull: true,
    },
    event_start: {
        type: DataTypes.DATE,
        allowNull: false
    },
    event_end: {
        type: DataTypes.DATE,
        allowNull: false
    },
    event_ticket_price: {
        type: DataTypes.DECIMAL(12, 2),
        defaultValue: 0,
    },
    event_type: {
        type: DataTypes.STRING(100),
        allowNull: true,
        comment: 'Loại sự kiện: concert, exhibition, festival, workshop...'
    },
    event_organizer: {
        type: DataTypes.STRING(255),
        allowNull: true,
    },
    event_capacity: {
        type: DataTypes.INTEGER,
        allowNull: true,
    },
    event_tags: {
        type: DataTypes.ARRAY(DataTypes.TEXT),
        defaultValue: []
    },
    images: {
        type: DataTypes.ARRAY(DataTypes.STRING),
        defaultValue: []
    },
    contact_info: {
        type: DataTypes.JSONB,
        allowNull: true,
    },
    total_attendees: {
        type: DataTypes.INTEGER,
        defaultValue: 0,
    },
    total_likes: {
        type: DataTypes.INTEGER,
        defaultValue: 0,
    },
    rating: {
        type: DataTypes.FLOAT,
        defaultValue: 0,
    },
    total_reviews: {
        type: DataTypes.INTEGER,
        defaultValue: 0,
    },
    is_active: {
        type: DataTypes.BOOLEAN,
        defaultValue: true,
    },
    is_featured: {
        type: DataTypes.BOOLEAN,
        defaultValue: false,
    },
}, {
    tableName: 'events',
    timestamps: true,
    underscored: true,
    indexes: [
        {
            fields: ['event_start', 'event_end']
        },
        {
            fields: ['lat', 'lng']
        }
    ]
});

export default Event;

